### PR TITLE
CFP for delegated IPAM with cilium-agent IPs

### PR DESCRIPTION
**Pull request by wedaly**
I started this draft CFP and prototype for getting delegated IPAM to work with endpoint health checking and ingress controller back in November. Unfortunately, I'm not going to have time to work further on this, so I'm publishing it as a draft PR to share what I found.

### CODE REVIEW

**Review by joestringer — COMMENTED** [review]
Thanks for sharing! I'll cc @squeed for awareness here. Notably just for the use case in general, and more specifically around the CNI spec interactions, GC operations and just general IPAM interaction.

I'm not doing a detailed review, but I did wonder about a couple of things while reading through the proposal.

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+The protocol for cilium-agent to call delegated IPAM is then relatively simple:
+
+1. If there is an IP to restore, invoke `CNI CHECK` to ensure that the IP is still allocated. If `CNI CHECK` succeeds, then return success.
+2. `CNI DEL` to ensure any previously-allocated IP is released. Continue to step 3 even if `CNI DEL` errors.
+3. `CNI ADD` to allocate a new IP. If it succeeds, then use the returned IP; otherwise, return failure.
+
+
+### Complications and caveats
+
+* **CNI state**: Some IPAM plugins store state on-disk (example: host-local writes to files in /var/lib/cni/networks by default, but this can be overridden in the CNI config). These directories *must* be mounted read-write in the cilium-agent pod, otherwise IPs could be leaked or double-allocated. Since this depends on the specific delegated IPAM plugin used, the user must configure this in the Cilium chart using `extraHostPathMounts`.
```
[the remark this answers is not in the document]
Comment (wedaly) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:86]: yeah, this is messy. And -- even worse -- it's easy to misconfigure and not notice, because things will _seem_ to work until IPs leak.

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+* **CNI Spec 1.1 GC operation**: [CNI spec 1.1 introduces a new "GC" operation](https://github.com/containernetworking/cni/pull/1022). The idea is that the container runtime calls GC with a list of all known attachments, and the CNI plugin cleans up any attachments not in the list. The cleanup includes invoking delegated IPAM plugins to release IPs. This is a problem, since the container runtime won't know about IPs that cilium-agent allocated for itself by invoking the IPAM plugin directly. One possible solution would be for Cilium CNI's GC operation to inject IPs allocated by cilium-agent before Cilium CNI invokes the delegated IPAM plugin's GC. Unclear if this is allowed or forbidden by the CNI spec.
+
+* **CNI conflist installation**: cilium-agent needs to read the CNI conflist, which might not yet exist if it's installed by another daemonset (e.g. when Cilium is configured with `cni.install=false`). Easy thing to do is exit with an error, but it would be better to retry or watch the conflist directory.
```
[the remark this answers is not in the document]
Comment (wedaly) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:94]: The main difference is that if Cilium exits with error on startup, it could look to the user like Cilium is broken, even though waiting for the conflist is expected behavior.

Contrast that with what happens in the normal CNI flow: kubelet/containerd keep the node "NotReady" until the conflist file is created.

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+|------------|--------------------------------|----------------------------------------|------------------------------|
+| ADD        | Allocate an IP                 | CNI_CONTAINERID, CNI_NETNS, CNI_IFNAME | IPs (possibly IPv4 and IPv6) |
+| DEL        | Release an IP                  | CNI_CONTAINERID, CNI_IFNAME            | Success/failure              |
+| CHECK      | Verify that an IP is allocated | CNI_CONTAINERID, CNI_NETNS, CNI_IFNAME | Success/failure              |
+
+(The above table is highly simplified, see the [CNI spec](https://www.cni.dev/docs/spec) for full details.)
+
+The semantics of the above operations differ significantly from how other Cilium IPAM implementations work. In particular, Cilium's `ipam.IPAM` struct supports idempotent allocation of a specific IP using [AllocateIP](https://github.com/cilium/cilium/blob/70ae8d0ef536de807aab849291e5a68758cb8d4d/pkg/ipam/allocator.go#L47). This is used to restore IPs on cilium-agent restart, ensuring that the IP doesn't change and potentially disrupt the dataplane. This isn't possible with delegated IPAM, because:
```
[the remark this answers is not in the document]
Comment (wedaly) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:54]: hmm, that's a good question. The main point I was trying to make here is that it's challenging to shoehorn delegated IPAM into the existing IPAM allocator interface. Maybe there's a way to implement `Allocate(ip)` as a CNI CHECK operation, but I'm not sure what arguments cilium-agent would send to the IPAM plugin (maybe there's some mapping from IP address to containerID/netNS/iface?)

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (joestringer) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:86]: Yeah. I think that this is the sort of thing that CNI 1.1's GC operation is intended to assist with, at least to mitigate the worst case scenarios. 

As a side-note, this also raises the point that well-planned lifecycle testing for this feature would help to also mitigate these corner cases.

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+* **Cilium CNI version**: Current default Cilium CNI version is 0.3.1, but the `CNI CHECK` operation isn't supported until 0.4.0. The Cilium CNI code is compatible with 0.4.0, so I think it's safe to set 0.4.0 in the conflist.
+
+* **CNI Spec 1.1 GC operation**: [CNI spec 1.1 introduces a new "GC" operation](https://github.com/containernetworking/cni/pull/1022). The idea is that the container runtime calls GC with a list of all known attachments, and the CNI plugin cleans up any attachments not in the list. The cleanup includes invoking delegated IPAM plugins to release IPs. This is a problem, since the container runtime won't know about IPs that cilium-agent allocated for itself by invoking the IPAM plugin directly. One possible solution would be for Cilium CNI's GC operation to inject IPs allocated by cilium-agent before Cilium CNI invokes the delegated IPAM plugin's GC. Unclear if this is allowed or forbidden by the CNI spec.
```
Comment (joestringer) [inline cilium/CFP-TODO-delegated-ipam-cilium-ips.md:92]: One more nit here, looks like GC doesn't arrive until 1.2. The linked PR's commit isn't tagged in 1.1 and the CNI webpages don't publish anything about the GC operation.

**Comment by squeed** [comment]
(apologies, I've been totally slammed, but reviewing this is very high up on my to-do list)

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+| CHECK      | Verify that an IP is allocated | CNI_CONTAINERID, CNI_NETNS, CNI_IFNAME | Success/failure              |
+
+(The above table is highly simplified, see the [CNI spec](https://www.cni.dev/docs/spec) for full details.)
+
+The semantics of the above operations differ significantly from how other Cilium IPAM implementations work. In particular, Cilium's `ipam.IPAM` struct supports idempotent allocation of a specific IP using [AllocateIP](https://github.com/cilium/cilium/blob/70ae8d0ef536de807aab849291e5a68758cb8d4d/pkg/ipam/allocator.go#L47). This is used to restore IPs on cilium-agent restart, ensuring that the IP doesn't change and potentially disrupt the dataplane. This isn't possible with delegated IPAM, because:
+
+* The required inputs do not include the IP address. By convention, some [IPAM plugins support an additional "ips" argument](https://www.cni.dev/docs/spec), but this is not universal.
+* The CNI ADD operation is not idempotent. According to [the spec](https://www.cni.dev/docs/spec/#add-add-container-to-network-or-apply-modifications): "A runtime should not call ADD twice (without an intervening DEL) for the same (`CNI_CONTAINERID`, `CNI_IFNAME`) tuple."
```
Comment (squeed) [inline cilium/CFP-TODO-delegated-ipam-cilium-ips.md:57]: CNI author here. That's why RFCs use "SHOULD" :-p.

But seriously, this is a good motivation to relax the no-add-without-del restriction. We've been idly discussing making ADD idempotent in CNI. I'll take this as another data point.

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md (hunk 4 above)
Comment (squeed) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:92]: No, this is my fault. The CNI repository is at v1.1, but that's the version of the go library. The spec version is v1.0.

I need to fix this. We will probably move the go library to its own repository.

File: cilium/CFP-TODO-delegated-ipam-cilium-ips.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (squeed) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:86]: Right, this is turning Cilium in to both a CNI plugin as well as a runtime. It is definitely complicated.

CNI plugins "expect" to be executed in the same context as the container runtime. Most notably, this means the real root filesystem, but it also means the same SELinux context. Now, I don't expect IPAM plugins to need significant privileges, but SELinux has tripped us up before.

This is a strong motivating example for a dumb "like CNI but gRPC" translation of the protocol; I'd been putting that off as not worth the effort, but this is a good motivating example.

There are a number of other potential stumbling blocks here, but I wont enumerate them all. It does mean that the administrator does have to have an intimate understanding of how the IPAM plugin is expected to run.

[the remark this answers is not in the document]
Comment (wedaly) [reply cilium/CFP-TODO-delegated-ipam-cilium-ips.md:86]: > CNI plugins "expect" to be executed in the same context as the container runtime. Most notably, this means the real root filesystem, but it also means the same SELinux context. Now, I don't expect IPAM plugins to need significant privileges, but SELinux has tripped us up before.

Ah, yes, I didn't even think about SELinux.

> It does mean that the administrator does have to have an intimate understanding of how the IPAM plugin is expected to run.

Totally agree. For AKS I'd expect things are simpler since we use an [IPAM plugin](https://github.com/Azure/azure-container-networking/blob/master/azure-ipam/ipam.go) that just calls an HTTP server without storing any state. But it's hard to solve the general case.

**Comment by squeed** [comment]
This is a clever solution, but I have concerns about the complicated dance that is Cilium acting as both a CNI plugin *and* a CNI runtime -- especially the details around lifecycle and making sure we don't leak IPs. CNI GC can save us, to some extent -- but only when runtime (containerd) implements it, which is in progress but not done yet :-).

What would it be like if health and ingress IPs were allocated by an external controller and set as annotations on the Node? Cilium could have a mode where it waits for these annotations before proceeding with startup.

Alternatively, what if we created a "placeholder" daemonset? Then there would be two noop pods scheduled to the node, but then the lifecycle is crystal clear.

**Comment by wedaly** [comment]
Thanks for reviewing this @squeed !

> This is a clever solution, but I have concerns about the complicated dance that is Cilium acting as both a CNI plugin and a CNI runtime -- especially the details around lifecycle and making sure we don't leak IPs. CNI GC can save us, to some extent -- but only when runtime (containerd) implements it, which is in progress but not done yet :-).

Yeah, after thinking this through I tend to agree. Even if CNI GC is able to clean up leaked IPs, I think there's a risk of double-allocating IPs if the volume mounts are misconfigured (cilium restarts, CNI IPAM state isn't preserved due to misconfigured volume mounts, and IPAM plugin returns an IP that was already assigned  to a pod). Not totally sure how Cilium handles two pods assigned the same IP.

> What would it be like if health and ingress IPs were allocated by an external controller and set as annotations on the Node? Cilium could have a mode where it waits for these annotations before proceeding with startup.

That seems feasible, although I'd worry about moving this responsibility to the cloud provider. The nice thing about delegated IPAM today is that Azure IPAM doesn't need to know that it's running Cilium, and Cilium doesn't need to know that it's using Azure IPAM.

In theory I guess someone could write a controller that tells Azure IPAM to allocate an IP, then annotates the nodes; however, there isn't an easy way to do this in AKS today. In most networking configurations, IPAM allocation happens by invoking an endpoint on a daemonset on the node where the IP is allocated; there isn't an API available that a controller could call to allocate an IP for any node.

> Alternatively, what if we created a "placeholder" daemonset? Then there would be two noop pods scheduled to the node, but then the lifecycle is crystal clear.

I like this approach better. No-op daemonset pods seem like a reasonable tradeoff to reduce the risk of misconfiguration.

I think there's a cyclic dependency to figure out though: (1) daemonset pod needs cilium-agent running to start and get an IP, (2) cilium-agent needs the IP from the daemonset pod to start cilium-agent. How hard would it be to let cilium-agent start without the IPs? For ingress/healthcheck IPs I guess cilium-agent could set these up later, but I'm less sure about the router IP.

**Comment by wedaly** [comment]
> In theory I guess someone could write a controller that tells Azure IPAM to allocate an IP, then annotates the nodes; however, there isn't an easy way to do this in AKS today.

I take that back: it is possible to assign additional IPs to the VM through Azure APIs, which is what Cilium's legacy Azure IPAM implementation does: https://github.com/cilium/cilium/blob/4820ebe7b372977c351837eb1ce5ac6024f712aa/pkg/azure/ipam/instances.go#L23

So that might be an option. It's a bit sad that it's cloud-provider specific, but at least someone could implement this controller outside the Cilium codebase.

> What would it be like if health and ingress IPs were allocated by an external controller and set as annotations on the Node?

One other quick thought on this: there's a security risk to granting a controller permission to update k8s nodes, since an attacker could manipulate labels to control where pods get scheduled (https://kccnceu2022.sched.com/event/ytlb/trampoline-pods-node-to-admin-privesc-built-into-popular-k8s-platforms-yuval-avrahami-shaul-ben-hai-palo-alto-networks). Maybe permission to mutate CiliumNode is safer?

**Comment by xmulligan** [comment]
@wedaly We now have status for the CFPs https://github.com/cilium/design-cfps#status. Would you consider moving this to dormant so it could be picked up again in the future?

**Comment by wedaly** [comment]
> @wedaly We now have status for the CFPs https://github.com/cilium/design-cfps#status. Would you consider moving this to dormant so it could be picked up again in the future?

Yep, that works for me. Where do I set the status?

**Review by joestringer — APPROVED** [review]
Similar to https://github.com/cilium/design-cfps/pull/7, I think that it's OK to merge dormant proposals in the state that they're in, even if there are remaining outstanding items to address (discussion also in https://github.com/cilium/design-cfps/issues/48). The CFP can be subsequently picked up by someone in the community resurrect it, and at that time they can make subsequent proposals to change the CFP and move the status to "Implementable". The proposal's acceptance would be subject to review at that time.

Prior to merge I wanted to take a fresh look across the proposal to try to give a high level take on where it's at right now, so that anyone attempting to pick this up again in future can get some guidance about what to address next. The comments above in this PR provide more specific feedback, but I'll try to summarize below.

I think the overall open question is whether we need more structure from the CNI spec (or successor) to formalize a solution. There is an alternative for a more limited scope, less clean solution through existing mechanisms. There's also the question of whether Cilium is assuming more of the role traditionally associated with the runtime rather than the CNI, and what hurdles that may introduce (including for instance, running in the host rather container context and being compatible with SELinux).

**Comment by joestringer** [comment]
To be merged into the repository in dormant state, we would need two more things:

1. Each commit needs to be signed off with DCO. You can see details and what this means in the "checks" section at the bottom of the PR.
2. Mark as "ready for review".

**Comment by wedaly** [comment]
> To be merged into the repository in dormant state, we would need two more things:
> 
>     1. Each commit needs to be signed off with DCO. You can see details and what this means in the "checks" section at the bottom of the PR.
> 
>     2. Mark as "ready for review".

thanks, updated!
